namespace/ga4gh-namespace-identifiers.md

```diff
@@ -0,0 +1,53 @@
+
+As a result, we are proposing that the computed identifier scheme described in the VR specification be considered for adoption as a GA4GH-wide standard. If the proposal is accepted by the GA4GH executive committee, the current VR proposal will stand as-is; if the proposal is rejected, the VR proposal will be modified to rescope the computed identifier mechanism to VR and under admininstration of the VR team.
+
+## Proposal
+
+The following algorithmic processes, described in depth in the VR [Computed Identifiers](https://vrs.ga4gh.org/en/1.0/impl-guide/computed_identifiers.html#computed-identifiers) proposal, are included in this proposal by reference:
+
+* **GA4GH Digest Serialization** is the process of converting an object to a canonical binary form based on JSON and inspired by similar (but unratified) JSON standards. This serialization for is used only for the purposes of computing a digest.
+
+* **GA4GH Truncated Digest** is a convention for using [SHA-512](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf), truncated to 24 bytes, and encoding using [base64url](https://datatracker.ietf.org/doc/html/rfc4648#section-5).
+
+* **GA4GH Identification** is the CURIE-based syntax for constructing a namespaced and typed identifier for an object.
+
+## Type Prefixes
+
+A GA4GH identifier is constructed according to this syntax:
+
+```
+"ga4gh" ":" type_prefix ":" digest
+```
+
+The `digest` is computed as described above. The type_prefix is a short alphanumeric code that corresponds to the type of object being represented. If this propsal is accepted, this “type prefix map” would be administered by GA4GH. (Currently, this map is maintained in a YAML file within the vr-spec repository, but it would be relocated on approval of this proposal.)
+
+We propose the following guidelines for type prefixes:
+
+* Prefixes SHOULD be short, ideally 2-4 characters.
+* Prefixes SHOULD be for concrete types, not polymorphic parent classes.
+* A prefix MUST map 1:1 with a schema type.
+
+See the [type prefix registry](./type-prefix-registry.md) for the current list of approved type prefixes under the `ga4gh` namespace.
+
+## Type Prefix Submission Process
+
+All GA4GH work streams are encouraged to use existing type prefixes in their specifications wherever possible, and to submit new type prefixes for data models that aren't currently represented in the registry.
+
+To add an entry to the type prefix registry:
+* Fork the TASC repository in your own user space
+* In your user space, create a new branch from `ga4gh:master`
+* In the new branch, add an additional row in the markdown table in `type-prefix-registry.md` with your proposed type prefix.
```

Comment:
In my opinion, the authoritative table should be code, preferably YAML, and not merely in documentation. That code should probably be in a new central/shared/core repo, e.g., `ga4gh-common`. PRs to add prefixes should be in that code. And, the documentation for those should be in the same repo, adjacent to the changes.

When I originally implemented prefixes for VRS, I anticipated separating GA4GH shared components from VRS. For this reason, the VRS repo has a separate file [ga4gh.yaml](https://github.com/ga4gh/vrs/blob/main/schema/ga4gh.yaml).

So, the most straightforward action would be to move this file to another repo and VRS clients would include that (as a subrepo or with vendoring).

Reply:
agreed. the new repo would be the authoritative GA4GH namespace registry.

Reply:
+1

Reply:
I also strongly agree.

Reply:
@reece do you have any examples of how the registry would be used in a machine-readable sense? We discussed this at TASC and saw it mainly as human-readable documentation. i.e. developers of data platforms/systems would read through the registry and find GA4GH prefixes related to data models they are using, and then apply our identifier system to their objects.

Reply:
@reece @larrybabb what's the purpose of creating a new repo? The `TASC` repo already fulfills one of the roles you describe (e.g. a cross-workstream registry for service info). Is it mainly that the name `TASC` doesn't describe the contents of the repo? If so, would it be sufficient to rename it?

Reply:
@jb-adams not that I want to speak for @reece or @larrybabb, but I do think the intent here is to be descriptive of the repo contents, as you suggest.

It might make sense to rename this repo to do so, but I think that there are several other TASC-related concerns that are tracked here; do you want to limit the scope of the TASC repo to concern itself only with the `ga4gh-common` technical product?

I think @reece sees such a repo to eventually serve as a complete replacement for the [GA4GH YAML in VRS](https://github.com/ga4gh/vrs/blob/main/schema/ga4gh.yaml) in addition to other technical artifacts to be coordinated / blessed centrally by GA4GH.

Reply:
@jb-adams : Apologies for missing the questions two weeks ago.

As @ahwagner suggested, please see the VRS repo.  The file he pointed to (ga4gh.yaml) is used to build a map of type ⇒ prefix, which is then used during identifier assignment. That is, when an object of type Allele is serialized, the prefix for that identifier is specified in the file ga4gh.yaml. In an ideal world, this mapping would be stored in exactly one place and all code *and documentation* would be derived from it.

Reply:
@jb-adams : IMO, "TASC" is a group and doesn't help understand what this code is about. "config" or "globals" might be a better name. Admittedly, I know of only one file that would go there now.

Despite being a primarily Python guy these days, I do think that files of this sort should be language neutral.

Reply:
Re repository naming: This is trivia. It doesn't matter what repo this prefix registry is in; what matters is that it exists. This repository currently contains the registry of Service Info types, and the PR proposes to add a registry of prefixes. Both of these are administered by the TASC committee, and IMHO this ga4gh/TASC repository is a convenient place to gather them.

The place that this information ought to be located in is a GA4GH technical web site. To date, GA4GH as an organisation has not provided a technical web site for the information produced by its working groups. The rumblings are that this is finally on the way to changing. In the meantime, this repository with the PR as proposed is as good a staging area as any; IMHO “keep it simple” applies.

Reply:
Re maps of type ⇒ prefix: VRS's _ga4gh.yaml_ file contains entries like `Allele: VA`. But what does `Allele` **mean**?

In the VRS context, it specifically means VRS's `Allele` object as defined in your schema. That's good and concrete.

However in the (lack of) context of this general registry list, it's just a word. Different groups will inevitably mean slightly different things when they use the word. That's why IMHO this registry's entries need to consist of the reserved prefix (`VA`), a prose description of the intention (“An allele, encompassing blah blah”), and links to its specific incarnations in individual GA4GH protocols (link to VRS's `Allele` initially; eventually also links to other groups' protocols' uses of `VA`).

I guess this is the crux of why I don't see this registry as a machine-readable list. What would the machine-readable entries contain? We are trying to make a list of registered prefixes like `VA` (and collate their use in GA4GH formats and protocols); we are not trying to define an ontology of terms like `Allele`. It's great to have a type/prefix map in a concrete context like VRS, but I don't see that it's feasible or useful for TASC to attempt in the general context.